File: src/Model/Statement/CloseRFileStatement.java
package Model.Statement;

import DataTypes.*;
import Model.Expression.Expression;
import Model.ProgramState;
import Exception.MyStatementExecutionException;
import Exception.ExpressionEvaluationException;
import Exception.InvalidKeyMyDictionaryException;

import java.io.BufferedReader;
import java.io.IOException;

/**
 * Created by vladc on 07.11.2016.
 */
public class CloseRFileStatement implements IStatement {
    private Expression var_file_id;

    public CloseRFileStatement(Expression var_file_id) {
        this.var_file_id = var_file_id;
    }

    public ProgramState execute(ProgramState state) throws MyStatementExecutionException {
        MyIDictionary<Integer, MyPair<String, BufferedReader>> fileTable = state.getFileTable();
        MyIDictionary<String, Integer> symTable = state.getSymTable();
        MyIHeap heap = state.getHeap();

        // 1. evaluate var_file_id to a value. If any error occurs then terminate the execution.
        int value;
        try {
            value = var_file_id.evaluate(symTable, heap);
        }
        catch (ExpressionEvaluationException ex) {
            throw new MyStatementExecutionException("The file is not opened for reading.");
        }

        // 2. Use the value to get the entry into the FileTable and get the associated BufferedReader
        // object. If there is not any entry in FileTable for the value we stop the execution.
        BufferedReader br;
        try {
            br = fileTable.get(value).getSecond();
        }
        catch (InvalidKeyMyDictionaryException ex) {
            // todo: choose an appropriate error message
            throw new MyStatementExecutionException("Error...");
        }

        // 3. call the close method of the BufferedReader object
        try {
            br.close();
        }
        catch (IOException ex) {
            throw new MyStatementExecutionException("IOException: " + ex.toString());
        }

        // 4. delete the entry from the FileTable
        try {
            fileTable.remove(value);
        }
        catch (InvalidKeyMyDictionaryException ex) {
            throw new MyStatementExecutionException("There is no entry (" + value + ") in the file table.");
        }

        // 5. return
        return state;
    }

    @Override
    public String toString() {
        return "closeRFile(" + var_file_id.toString() + ")";
    }
}

File: Toy-Language-Interpreter/Model/Statement/CloseRFileStatement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Toy_Language_Interpreter.DataTypes;
using Toy_Language_Interpreter.Model.Expression;
using Toy_Language_Interpreter.MyException;

using System.IO;

namespace Toy_Language_Interpreter.Model.Statement
{
    class CloseRFileStatement : IStatement
    {
        private AExpression var_file_id;

        public CloseRFileStatement(AExpression var_file_id)
        {
            this.var_file_id = var_file_id;
        }

        public ProgramState Execute(ProgramState state)
        {
            MyIDictionary<int, MyPair<string, StreamReader>> fileTable = state.GetFileTab();
            MyIDictionary<string, int> symTable = state.GetSymTab();

            // 1. evaluate var_file_id to a value. If any error occurs, stop the execution
            int value;
            try
            {
                value = var_file_id.Evaluate(symTable);
            }
            catch (ExpressionEvaluationException e)
            {
                throw new MyStatementExecutionException("The file is not opened for reading.");
            }

            // 2. Use the value to get the entry into the file table and get the associated StreamReader
            // object. If there is no entry in the file table for the value, we stop the execution
            StreamReader sr;
            try
            {
                sr = fileTable.Get(value).GetSecond();
            }
            catch (InvalidKeyMyDictionaryException)
            {
                throw new MyStatementExecutionException("Error. There is no entry (" + value + ") in the file table.");
            }

            // 3. call the Close() method
            try
            {
                sr.Close();
            }
            catch (IOException e)
            {
                throw new MyStatementExecutionException("IO Error: " + e.ToString());
            }

            // 4. delete the entry from the file table
            try
            {
                fileTable.Remove(value);
            }
            catch (InvalidKeyMyDictionaryException)
            {
                throw new MyStatementExecutionException("Error. There is no entry (" + value + ") in the file table.");
            }

            // 5. return
            return null;
        }

        public override string ToString()
        {
            return "closeRFile(" + var_file_id + ")";
        }
    }
}
